README.md
# blindController
Jeenode / Arduino Roller Blind Controller - reports via RF to emonhub (Openenergymonitor), remote control via RF, intelligent opening /closing based on light, sun dipping and more.

blindController.pl / blindController.cfg - this is a configuration program to allow you set various paramaters on the blinds (light open/close thresholds, timers, overrun parameters, operation mode and more).  Uses a mysql database to hold the config for all your blinds and send to them at will. blindController.sql holds the SQL needed to create the database.

Also has the ability to read data from a WH1080 weather station and relay it in Jeelib RF packet format.
Finally, a DS18B20 temperature sensor can be connected across the 'OPEN' switch and the code will relay temperature back to the base station too.

If using emonhub, here is the config line you will need to receive each blinds status data, and send data to the blind :

[[2]]
    nodename = example_blind
   firmware =V1_4
   hardware = blindControllerv3
    [[[rx]]]
     names = dest,opMode,bsm0,bsm1,tom,ldrval,temp,position0,position1,fwVersion
     datacodes = b,b,h,h,h,h,h,f,f,h
     scales = 1,1,1,1,1,1,0.1,1,1,1
     units = id,opmode,bsm,bsm,tsm,ldr,temp,pos,pos,ver
    [[[tx]]]
     names = cmd,opmode,nodeid,autoOpenMin,autoCloseMin,sunDipSec,sunDipPosition,openThreshold,closeThreshold,sundiptThreshold,autoWaitMin,manWaitMin,opMode,overRun,underRun,crc
     datacodes =b,b,b,b,b,h,b,h,h,h,b,b,b,b,b,H
     units = cmd,opmode,min,min,sec,pct,thresh,thresh,thresh,min,min,opmode,pct,pct,crc



Enjoy.

alandpearson / july 2022


CHANGES
BlindController for Roman or Roller blinds using a JeeNode, LDR, DC motor & limit microswitches

Works by assessing reading from LDR and driving blinds accordingly. SunDip feature will partially close the blinds if it it too bright for longer than SunDipThreshold time and open them again afterwards. Configurable via RF. Different modes of operation possible from manual to fully automated.

(c) 2018 alandpearson@gmail.com

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.


-------------------------------------------------------------------------------------------
v1.4
Many changes including adding the version number in the code and reporting via RF.
Added WH1080 weather station reporting (optional compile time DEFINE)
Added DS18b20 temp sensor support


v1.3

blindPosition saved to NVRAM every time a move operation completes meaning that if power is lost blind stays in same position on power restore.
If blind was in middle of moving during power loss, then blind will move to open position on power restoration

sunDip now exits when < closeThreshold (was < openThreshold)
Also sunDip exit can now go directly to closed position if in the unlikely situation you go from sundip brightness to closing darkness without any in between

powerOn Reset now exists and will wipe saved blindPosition from NVRAM.
to use, start blind, remove power within 30 seconds and power up again.
(the POR flag set in setup and is reset when loop() has been running more that 30 secs).

overrun (bottom limit miss)  and underrun (top limit miss) values lowered to 1% and 3% respectively and are part of blindConfig

v1.2
Calibration data saved to NVRAM meaning will only recalibrate when requested by hold open+close buttons at startup.

Default behaviour is to open blind at power on

v1.1
I don't remember ?
Maybe add RF reporting and control via emonpi ?


v1.0
A long time ago,my memory since faded
Something that worked... 





wh1050.h
/// @dir wh1050_relay
///2020-12-24 alandpearson@gmail.com
/// Based on the feneralized decoder and relay for 868 MHz and 433 MHz OOK signals.
// 2010-04-11 <jc@wippler.nl> http://opensource.org/licenses/mit-license.php




// RF12 communication settings
//Set the RFM12B to Jeenode OOK RSI connection
#define OOK_INT 3    //3 = D3 = ext int 1 = JeeNode P1-4 IRQ
//4 = D4 = JeeNode P1 - DIO
//7 = D7 = JeeNode P4 - DIO
//14 = A0/D14 = Jeemode P1 - AIO
//17 = A3/D17 = Jeemode P4 - AIO

#include "decoders.h"
#include "rf12_helper.h"


// I/O pin allocations, leave any of these undefined to omit the code for it
//NB if using RFM12 as OOk receiver, it needs to be connected to an Arduino pin
//So you must also use PIN_433 to tell the code which pin the RFM12 is connected to


#define SYNC_EFFORTS 2  // during sync, how many times we need to receive a broadcast from same WH1050 Id

#define SYNC_TIMEOUT_SECS 600 // We must see the same WH1050 Id in this time, or we will try to resync
//The below settings are important. The WH1050 will send data every 48 secs
//Ideal is in you catch that window on the nail, but that isn't always possible
//So what you want to do is turn on OOK a few secs before we expect the packet
//and run in OOK mode for a few secs after we would expect it
#define OOK_ON_DURATION 3 // Secs we are in OOK receive mode (maximum) - Will be less if we recieve a packet


//WX_SYNC_LISTEN_SECS
//The max time we can stay in OOK listening for a Weather station
//This will be applied on a hourly basis
//This has the affect if we cannot find a station in WX_SYNC_LISTEN_SECS
//then we must wait a full hour to try to sync again
//This value should not be too big, or code that uses this library
//will not be able to send/receive in FSK mode for the duration
#define WX_SYNC_LISTEN_SECS 180  //Maximum time we can be looking for a weather station while syncing (seconds)
//We don't want this too long as we will be in OOK mode all this time,
//meaning FSK reception not possible and interrupts going mad, maybe screwing blind timers.

bool wxReady = false ;        // If data has been received via OOK Decoder and ready to be processed
static uint8_t wxPacket[9];
uint32_t wxLastHeard = 0 ;    // Last millis() we heard a broadcast from any wx station
uint32_t mywxLastHeard = 0 ; // Last millis() we heard a broadcast from wx station we are synced to

static uint8_t prev_rssi = 0; //last packet width
uint32_t  lastRtcUpdate = 0 ; // Last millis() we heard an RTC packet via RF network
uint32_t fskOnTime = 0 ;      //Time we switched RF12 to FSK
uint32_t ookOnTime = 0 ;      //Time we switched RF12 to OOK
DateTime dcfData ;            //Struct to hold the DCF data to be sent via OOK

enum rxModes { FSK, OOK };
rxModes rxMode ;


WH1080DecoderV2a wh1080v2a_9(9);
PayloadRTC rtcData ;


//Do we have onbaord BMP280 pressure sensor ?
#ifdef BMP280
#include <BMP280_DEV.h>                           // Include the BMP280_DEV.h library
BMP280_DEV bmp280;                                // Instantiate (create) a BMP280_DEV object and set-up for I2C operation
#endif




#ifdef DEBUG

void reportSerial (const char* s, class DecodeOOK& decoder) {
  byte pos;
  const byte* data = decoder.getData(pos);
  Serial.print(s);
  Serial.print(":[");
  for (byte i = 0; i < pos; ++i) {
    //Serial.print(data[i] >> 4, HEX);
    //Serial.print(data[i] & 0x0F, HEX);
    Serial.print(data[i], HEX) ;
    Serial.print(':') ;
  }
  Serial.println(']');
}

#endif

void sendWxData(wh1050Payload data) {

  if (rxMode == FSK ) {
    rf12_initialize(WXNODEID, RF12_433MHZ, NETGRP);
    delay(500) ;
    while (!rf12_canSend())
      rf12_recvDone();
    rf12_sendStart(WXNODEID, &data, sizeof data);
    rf12_sendWait(1);
    Serial.println(F("Sent RF WX reading")) ;
  } else {
    Serial.println(F("Cant RF send - not in FSK mode!")) ;
  }

  rf12_initialize(MYNODEID, RF12_433MHZ, NETGRP);


}


void fskReceiveOn() {

  if (rxMode != FSK) {
    rf12_stop_rx_OOK();
    rf12_initialize(MYNODEID, RF12_433MHZ, NETGRP);
#ifdef DEBUG
    Serial.println(F("FSK mode")) ;
    delay(500) ;
#endif

    rxMode = FSK ;
    fskOnTime = millis() ;
  }
}

void ookReceiveOn() {

  if (rxMode != OOK) {
    //  rf12_initialize(NODEID, RF12_433MHZ, NETGRP);
    rf12_initialize(0, RF12_433MHZ, 0);

    delay (250) ;
    rf12_start_rx_OOK() ;
#ifdef DEBUG
    Serial.println(F("OOK mode")) ;
#endif

    rxMode = OOK ;
    ookOnTime = millis() ;
  }
}



bool pollOOK() {
  // Read the RSSI value from the last interrupt
  blockInterrupts();
  unsigned long rssi_diff = rf12_pulse;
  unsigned long rssitime = rf12_tstamp;
  uint8_t rssi = rf12_signal;
  allowInterrupts();

  if (rssi != prev_rssi) {

    if (rssi_diff > 32000)
      rssi_diff = 32000;

    if (wh1080v2a_9.nextPulse(rssi_diff, prev_rssi) ) {
      byte num_bytes = 0;
      const uint8_t* bytes = wh1080v2a_9.getData(num_bytes);
      memcpy(wxPacket, bytes, num_bytes);
#ifdef DEBUG
      reportSerial("WH1080DecoderV2a", wh1080v2a_9); //Note: clears part of data buffer!
#endif
      wh1080v2a_9.resetDecoder();
      wxLastHeard = millis() ;

      wxReady = 1;
    }

  }

  prev_rssi = rssi;

  return (wxReady) ;

}



bool wxSync(bool resync = 0 ) {


  static byte count = 0 ;
  static byte wxStnID = 0, lastwxStnID = 0 ;
  static byte synced = 0 ;
  static byte syncing = 0 ;
  static uint32_t nextSyncAttempt = 0 ;
  static uint32_t nextOOKonTime = 0 ;



  if ( (millis() - mywxLastHeard) > (SYNC_TIMEOUT_SECS * 1000ul) ) {
    synced = 0 ;
  }

  if (! synced && millis() > nextSyncAttempt || resync ) {

    //Get syncing !

    if (resync) {
      //reset the last ookOnTime timestamp
      //by entering FSK then OOK mode
      fskReceiveOn() ;
    }

    ookReceiveOn() ;

    if (millis() - ookOnTime >  (WX_SYNC_LISTEN_SECS * 1000ul) ) {
      //We've been in syncing too long and not found a Weather Station
      //Give up for another hour
      nextSyncAttempt = millis() + 3600000ul ;
      nextOOKonTime = nextSyncAttempt ;
      Serial.println(F("wxSync: Give up sync for 1 hr")) ;
      fskReceiveOn() ;
    } else {
      //Do the sync dance

      if (!synced && !syncing) {
        Serial.println(F("wxSync: Resyncing")) ;
        count = 0 ;
      }

      if ( pollOOK() && (count < SYNC_EFFORTS) ) {

        //only do this so we can display the decoded packet if in DEBUG mode
        wh1050Payload wxReading = decodeSensorData(wxPacket) ;

        wxStnID = wxReading.stnID ;
        Serial.print(F("Sync: heard Wx stn ID:")) ; Serial.println(wxStnID, HEX);

        if ( wxStnID == lastwxStnID ) {
          lastwxStnID = wxStnID ;
          mywxLastHeard = millis() ;
          count ++ ;
        } else {
          lastwxStnID = wxStnID ;
          count = 1;
        }

        wxReady = 0;
      }

      // Have we got a valid (same) station ID at least 3 times in a row ?

      if (count >= SYNC_EFFORTS && wxStnID == lastwxStnID ) {
        // good - we found a station SYNC_EFFORTS times in a row, set the global var to be this stations ID
        myStnId = wxStnID ;
        if ( count == SYNC_EFFORTS) {
          //only print this message once per sync
          Serial.print(F("Synced to Wx stn ID:")) ; Serial.println(wxStnID, HEX) ;
          count ++;
        }
        synced = 1;
        syncing = 0 ;
      } else {
        synced = 0 ;
        syncing = 1;
      }

    }

  } else {
    //Synced or not trying to sync.
    //Manage going in and out of OOK mode
    //every 48 seconds for OOK_ON_DURATION
    //Turn OOK mode on 48secs - OOK_ON_DURATION / 2
    //to get OOK each side of the window we expect the packet

    //if we find OOK has been on for more than OOK_ON_DURATION
    //turn it off. In this case, we did not receive a packet
    //as expected, so recalculate the next expected time

    //only turn OOK on at nextOOKonTime (avoids going on/off when not synced and giving up for 1 hour above)

    //The next time we expect a packet
    //As OOK_ON_DURATION is in seconds, multiply *500 will give us half the value in millisecs
    if (millis() - wxLastHeard < 48000  && nextOOKonTime < millis() ) {
      nextOOKonTime = wxLastHeard + ( 48000ul - (OOK_ON_DURATION * 500ul) ) ;
    } else if ( nextOOKonTime < millis() ) {
      nextOOKonTime = nextOOKonTime + (48000ul) ;
      //do not add on the OOK_ON_DURATION as this has been factored in
      //above. Otherwise we would drift by OOK_ON_DURATION/2 everytime a packet is missed
    }


    //Make sure we go back into FSK mode if OOK has been on too long

    if ( rxMode == OOK && millis() > ookOnTime + (OOK_ON_DURATION * 1000ul) ) {
      fskReceiveOn() ;
    } else if (rxMode == FSK && millis() >= nextOOKonTime) {
      ookReceiveOn() ;
    }

  }

  return (synced) ;

}



void wxSetup() {
  Serial.println(F("\nwxRelay start up.."));



  //Start the emulated RTCClock
  RTC_Millis::begin (&dcfData) ;



#ifdef BMP280

#ifdef DEBUG
  Serial.println(F("BMX280 sensor setup")) ;
#endif

  bmp280.begin(BMP280_I2C_ALT_ADDR);      // Default initialisation, place the BMP280 into SLEEP_MODE
  bmp280.setTimeStandby(TIME_STANDBY_4000MS);     // Set the standby time to 4 seconds
  bmp280.startNormalConversion();                 // Start BMP280 continuous conversion in NORMAL_MODE

#ifdef DEBUG
  Serial.println(F("BMX280 setup done")) ;
#endif


#endif

  ookReceiveOn() ;


  Serial.println(F("wxRelay started")) ;
}


byte wxLoop() {

  byte state = 0 ; //used as return value so caller can see our state


  //Find a wh1050 weather station and keep us in sync with it
  if ( wxSync(0) ) {
    state = 2;

    //If we received wh1050 data (received & decoded successfully)
    if ( pollOOK() ) {


      //put the nice data into a wh1050Payload struct
      wh1050Payload wxReading = decodeSensorData(wxPacket) ;

#ifdef BMP280
      bmp280.getCurrentTempPres( wxReading.intempC, wxReading.hpa)  ;
#ifdef DEBUG
      Serial.print(wxReading.hpa); Serial.print(F(" hpa, ")) ;
      Serial.print(wxReading.intempC); Serial.println(F(" C"));
#endif
#endif

      wxReady = false ;


      if (wxReading.stnID == myStnId ) {

        state = 3 ;

        mywxLastHeard = wxLastHeard ;
        //bang out a DCF time sync packet if we received a RTC packet from emonhub in the last 24 hours
        //& minutes between 3 & 8 so as not to clog the airwaves
        dcfData = RTC_Millis::now () ;

        if (millis() < lastRtcUpdate + 86400000ul && lastRtcUpdate != 0 && (dcfData.minute() >= 3 && dcfData.minute() <= 8) ) {
          delay(500) ;

          sendDcfPacket( myStnId, dcfData );
#ifdef DEBUG
          Serial.println(F("Sent DCF data")) ;
#endif
          delay(300) ;

        }

        //Go into FSK mode immediately after packet RX
        fskReceiveOn() ;


        //Send Wx packet out via RFM12
        sendWxData(wxReading) ;

      } else {
#ifdef DEBUG
        //Heard another wx station - don't update mywxLastHeard
        Serial.println(F("Got Wx data for ID not synced to - ignored")) ;
#endif
      }


    } //if (pollOOK())


  } else {

    //Have not found a Wx Station
    state = 1 ;

  }


  return (state) ;


}


rf12_helper.h
#define STATUS_RSSI    (1 << 8)

#if OOK_INT == 3
#define PINCHGINT_OOK 0
#define RF12_RX_EXTINT 1
#elif OOK_INT == 2
#define PINCHGINT_OOK 0
#define RF12_RX_EXTINT 0
#else
#define PINCHGINT_OOK 1
#define RF12_RX_EXTINT 1
#endif

//#define RF12_RX_DATA 3  //D3 = bit 3 in PORTD - JeeNode!
//#define RF12_RX_EXTINT 1

#if OOK_INT < 8
#define RF12_RX_DATA OOK_INT
#define RF12_RX_PIN PIND
//#define RF12_RX_DDR DDRD
#elif OOK_INT < 14
#define RF12_RX_DATA OOK_INT - 8
#define RF12_RX_PIN PINB
//#define RF12_RX_DDR DDRB
#else
#define RF12_RX_DATA OOK_INT - 14
#define RF12_RX_PIN PINC
//#define RF12_RX_DDR DDRC
#endif


volatile static unsigned long   rf12_pulse;
volatile static unsigned long   rf12_signal;
volatile static unsigned long   rf12_tstamp;
static unsigned long            prev_tstamp = 0;
long rf12_dur_sum = 0;
long rf12_flip_count = 0;
long rf12_high_count = 0;
long rf12_low_count = 0;
long rf12_short_count = 0;
long rf12_int_dur = 0;




//PayloadRTC is the same struct that emonGLCD uses for its power 
//and happens to contain a clock update, so we'll sniff it and use it
//for our own RTC too.
typedef struct {
  byte id ;
  byte type ;     //should always be 0A - EmonPi Power Packet
  byte  hour, min, sec, day, month, year ;
  int utilityW, solarW, utilityKwh, solarKwh;
} PayloadRTC;   



static void blockInterrupts () {
#if PINCHGINT_OOK
#if OOK_INT < 8
  bitClear(EIMSK, PCIE2);
#elif OOK_INT < 14
  bitClear(EIMSK, PCIE0);
#else
  bitClear(EIMSK, PCIE1);
#endif
#else
  bitClear(EIMSK, RF12_RX_EXTINT);
#endif
}

static void allowInterrupts () {
#if PINCHGINT_OOK
#if OOK_INT < 8
  bitSet(EIMSK, PCIE2);
#elif OOK_INT < 14
  bitSet(EIMSK, PCIE0);
#else
  bitSet(EIMSK, PCIE1);
#endif
#else
  bitSet(EIMSK, RF12_RX_EXTINT);
#endif
}



#if OOK_INT >= 14
//#define VECT PCINT1_vect
#define VECT ANALOG_COMP_vect
#elif OOK_INT >= 8
#define VECT PCINT0_vect
#else
#define VECT PCINT2_vect
#endif



void rf12_data_int(void) {
  rf12_signal = bitRead(RF12_RX_PIN, RF12_RX_DATA);
  rf12_tstamp = micros();
  // determine the pulse length in microseconds, for either polarity
  rf12_pulse = rf12_tstamp - prev_tstamp;
  prev_tstamp = rf12_tstamp;
  if (rf12_signal) {
    rf12_high_count++;
    //Serial.print(rf12_pulse);
    //Serial.print("-");
  } else {
    rf12_low_count++;
    //Serial.println(rf12_pulse);
  }
  rf12_flip_count++;
  rf12_dur_sum += rf12_pulse;
  if (rf12_pulse < 80)
  {
    rf12_short_count++;
  }
  rf12_int_dur += micros() - prev_tstamp;
  //Serial.println("rf12_data");
}

#if PINCHGINT_OOK
ISR(VECT) {
  rf12_data_int();
}
#endif

void rf12_interruptcontrol(byte attach) {
#if PINCHGINT_OOK
#if OOK_INT < 8
  if (attach) {
    bitClear(DDRD, OOK_INT);      // input
    bitSet(PORTD, OOK_INT);       // pull-up
    bitSet(PCMSK2, OOK_INT);      // pin-change
    bitSet(PCICR, PCIE2);         // enable
  } else
    bitClear(PCMSK2, OOK_INT);
#elif OOK_INT < 14
  if (attach) {
    bitClear(DDRB, OOK_INT - 8);  // input
    bitSet(PORTB, OOK_INT - 8);   // pull-up
    bitSet(PCMSK0, OOK_INT - 8);  // pin-change
    bitSet(PCICR, PCIE0);         // enable
  } else
    bitClear(PCMSK0, OOK_INT - 8);
#else
  if (attach) {
    bitClear(DDRC, OOK_INT - 14); // input
    bitSet(PORTC, OOK_INT - 14);  // pull-up
    //bitSet(PCMSK1, OOK_INT - 14); // pin-change
    //bitSet(PCICR, PCIE1);         // enable

    // enable analog comparator with fixed voltage reference
    ACSR = _BV(ACBG) | _BV(ACI) | _BV(ACIE);
    ADCSRA &= ~ _BV(ADEN);
    ADCSRB |= _BV(ACME);
    ADMUX = OOK_INT - 14;
  } else
    bitClear(PCMSK1, OOK_INT - 14);
#endif
#else

  if (attach) {
    attachInterrupt(RF12_RX_EXTINT, rf12_data_int, CHANGE);
  } else {
    detachInterrupt(RF12_RX_EXTINT);
  }
#endif
}

void rf12_start_rx_OOK()
{
  //   Serial.println("would be nice") ;
  prev_tstamp = micros();
  rf12_pulse = 0;
  rf12_signal = bitRead(RF12_RX_PIN, RF12_RX_DATA);
   
  rf12_dur_sum = 0;
  rf12_flip_count = 0;
  rf12_high_count = 0;
  rf12_low_count = 0;
  rf12_short_count = 0;
  rf12_int_dur = 0;

  rf12_control(0x8017); // 8027    868 Mhz;disabel tx register; disable RX
  //         fifo buffer; xtal cap 12pf, same as xmitter
  //rf12_control(0x82c0); // 82C0    enable receiver; enable basebandblock
  //rf12_control(0x82D8); // 82D8    enable receiver; enable basebandblock , crysal Osc + synth
  rf12_control(0x82F8); // 82F8    enable receiver & transmitter; enable basebandblock , crysal Osc + synth

  rf12_control(0xA620); // A620    433.92 MHz
  rf12_control(0xc691); // C691    c691 datarate 2395 kbps 0xc647 = 4.8kbps

//These setting are critical and determine the noise floor for OOK on the RFM
//If wrong then due to the damn capacitor documented here :
//the pulse widths can be all over the shop and OOK decode will not work.
//I've found the 97 + 91dbm to be the most successful

   rf12_control(0x9481); // 9481    VDI; FAST;200khz;GAIn Max; DRSSI -97dbm (too noisey) (best for AlOffice Jeenode)
// rf12_control(0x9482); //        VDI; FAST;200khz;GAIn Max; DRSSI -91dbm (too noisey - 30/12/2020 looks good) (best for USB Jeenode)
// rf12_control(0x9483); // 9483    VDI; FAST;200khz;GAIn Max; DRSSI -85dbm (looks good - 30/12/2020 a bit deaf)
// rf12_control(0x9484); // 9483    VDI; FAST;200khz;GAIn Max; DRSSI -79dbm (deaf)


  rf12_control(0xC220); // C220    datafiltercommand; ** not documented cmd
  rf12_control(0xCA00); // CA00    FiFo and resetmode cmd; FIFO fill disabeld
  rf12_control(0xC473); // C473    AFC run only once; enable AFC; enable
  //         frequency offset register; +3 -4
  rf12_control(0xCC67); // CC67    pll settings command
  rf12_control(0xB800); // TX register write command not used
  rf12_control(0xC800); // disable low dutycycle
  rf12_control(0xC040); // 1.66MHz,2.2V not used see 82c0
  
  rf12_interruptcontrol(1); //attach

}

void rf12_stop_rx_OOK()
{
  rf12_interruptcontrol(0); //detach
  rf12_dur_sum += micros() - prev_tstamp;
 
}




// Turn transmitter on or off, but also apply asymmetric correction and account
// for 25 us SPI overhead to end up with the proper on-the-air pulse widths.
// With thanks to JGJ Veken for his help in getting these values right.

static void ookPulse(int on, int off) {
  rf12_onOff(1);
  //  delayMicroseconds(on + 150);
  delayMicroseconds(on + 150);

  rf12_onOff(0);
  //delayMicroseconds(off - 200);

  delayMicroseconds(off - 150);

  }

static void fs20sendBits(word data, byte bits) {

  for (word mask = bit(bits - 1); mask != 0; mask >>= 1) {
    int width = data & mask ? 500 : 1500 ;

    ookPulse(width, 1000);
  }
}


static void sendPacket(uint8_t const * data, int len, uint8_t repeat = 1)
{

  while (repeat-- > 0)
  {

    uint16_t tx = 0;
    uint8_t b = 16;
    for (uint8_t i = 0; i < len; ++i)
    {
      fs20sendBits(data[i], 8) ;
    }



   delay(200);
  }
}


void sendDcfPacket( const uint8_t id,  DateTime &data )
{
  byte preamble = 0xFF ;
  uint8_t msg[10];

  msg[0] = preamble;
  msg[1] = (MsgType_DCF << 4) | (id >> 4);
  msg[2] = (id << 4) | 0x0a;
  msg[3] = dec2bcd(data.hour());// | 0x80;
  msg[4] = dec2bcd(data.minute());
  msg[5] = dec2bcd(data.second());
  msg[6] = dec2bcd(data.year() - 2000);
  msg[7] = dec2bcd(data.month());// | 0x40;
  msg[8] = dec2bcd(data.day());
  // WH1080 format is 11 bytes inc this : msg[9] = 0x45;
  msg[9] = crc8(&msg[1], 8);

  /*
    for (int i = 0; i < 15; i++) {
      msg[1] = 0x60 | 0xc ;
      Serial.print("  DCF ID:") ; Serial.println(msg[1], HEX) ;
      sendPacket(msg, ARRAY_SIZE(msg));

    }
  */
  //msg[1] = 0x6c  ;

  sendPacket(msg, ARRAY_SIZE(msg), 3);
}


void start_rx_OOK()
{
  rf12_start_rx_OOK();
}

void stop_rx_OOK()
{
  rf12_stop_rx_OOK();
}


blindConfigure.pl
#!/usr/bin/perl -w

#Configure blindControllers from database, sending via RF using mqtt to emonpi

use DBI;
use Net::MQTT::Simple;
use Net::MQTT::Simple::Auth;
use POSIX  'strftime';
use Config::Simple;

use File::Basename qw(dirname);
use Cwd  qw(abs_path);
use lib dirname(dirname abs_path $0) . '/libs/perl';

use HomeAutomation ;


my $cfg = new Config::Simple('blindConfigure.cfg') ;

#Database params
my $dbName = $cfg->param('mysql_db');
my $dbUser = $cfg->param('mysql_user');
my $dbPass = $cfg->param('mysql_pass');
my $dbHost = $cfg->param('mysql_db_hostname');

#MQTT params
my $mqttTopic = $cfg->param('mqtt_topic') ;

my $dbh = DBI->connect("DBI:mysql:$dbName:$dbHost", $dbUser, $dbPass);



sub menu {


	my $nodeId = 0 ;

	print "Blind Controller configuration programme\n\n";
	print "Found the following blind controllers in the database \n";
        print "\tnodeId: name\n";


	my $sth = $dbh->prepare("SELECT nodeId,name  FROM blindConfig");
	$sth->execute();

         while (@data = $sth->fetchrow_array()) {
	    $nodeId = $data[0];
            my $name = $data[1];
            print "\t$nodeId: $name\n";
          }

          if ($sth->rows == 0) {
            print "No blinds found in the database !\n\n";	
	    $nodeId = 0;
          } else {

		print "Enter nodeId you want to configure\n";
		$nodeId = <STDIN> ;
		chomp $nodeId ;

	  }

          $sth->finish;


	return $nodeId ;


}

sub mqttConfigMsg {

	my $nodeId = $_[0] ;

	print "nodeId" . $nodeId ;

}


sub printNodeConfig {
	
	my $nodeId = $_[0] ;
	my $mqttValues = "0" ;

	my $sth = $dbh->prepare("SELECT * FROM blindConfig where nodeId= $nodeId");
	$sth->execute();

          if ($sth->rows == 0) {
            print "Configuration not found for node $nodeId in the database !\n\n";	
	    $mqttValues = "0";
          } else {

		print "\n*****************************\n";
		print "Configuration for nodeId " . $nodeId . "\n" ;
		$ref = $sth->fetchrow_hashref ;

		$mqttValues="4," . 
			$ref->{'opMode'} . "," .
			$ref->{'nodeId'} . "," .
			$ref->{'autoOpenMin'} . "," .
			$ref->{'autoCloseMin'} . "," .
			$ref->{'sunDipSec'} . "," .
			$ref->{'sunDipPosition'} . "," .
			$ref->{'openThreshold'} . "," .
			$ref->{'closeThreshold'} . "," .
			$ref->{'sundipThreshold'} . "," .
			$ref->{'autoWaitMin'} . "," .
			$ref->{'manWaitMin'} . "," .
			$ref->{'opMode'} . "," .
			$ref->{'underRun'} . "," .  
			$ref->{'overRun'} ; 
			print "$mqttValues \n";

		foreach $key (keys %$ref)  {
		    print "${key}: " ;
		    print $ref->{$key},"\n"; 
		}

		print "*****************************\n";

		print "Proceed with sending configuration to node $nodeId ? \n";
		my $response = <STDIN> ;
		chomp $response ;
		if ( $response eq 'y' || $response eq 'Y') {
			print "Confirmed.\n";
			 #cmd,opmode,nodeid,autoOpenMin,autoCloseMin,sunDipSec,sunDipPosition,openThreshold,closeThreshold,sundipThreshold,autoWaitMin,manWaitMin,opMode,underRun,overRun,crc

		} else {
			#Cancelled
			$mqttValues = "1";
		}

	  }

	return $mqttValues ;


}




#static word calcCrc (const void* ptr, byte len) {
#  word crc = ~0;
#  for (byte $i = 0; $i < len; ++i)
#    crc = _crc16_update(crc, ((const byte*) ptr)[i]);
#  return crc;
#}



sub crc16_update {
   
	my ($crc, $a) = @_ ; 
	my $i = 0 ;

    $crc ^= $a;
    for ($i = 0; $i < 8; ++$i)
    {
        if ($crc & 1) {
            $crc = ($crc >> 1) ^ 0xA001;
       } else {
            $crc = ($crc >> 1);
	}
    }

    return $crc;
}



sub calcCrc {

	my $mqttMsg = $_[0] ;


	my $crc = 65535;
	my @values = split(',', $mqttMsg) ;
	print "$mqttMsg \n" ;
	$data = pack("CCCSCSSSCCCCC", $values[2], $values[3], $values[4], $values[5], $values[6] , $values[7], $values[8], $values[9], $values[10], $values[11], $values[12], $values[13], $values[14]);

	for ($i=0 ; $i < 17 ;  ++$i) {
		$crc = crc16_update($crc, vec $data,$i,8 ) ;
	}

	return ($crc) 
}





#########################################################################################################################
# Main															#
#########################################################################################################################

my $mqtt ; 
my $nodeId = 0 ;
my $mqttMsg = "0" ;
my %mqttData = () ;

# MQTT Connect 
if ($cfg->param('mqtt_enabled') eq "true") {
        print ("MQTT enabled \n");
        $mqtt = Net::MQTT::Simple::Auth->new($cfg->param('mqtt_hostname'), $cfg->param('mqtt_user'), $cfg->param('mqtt_pass') );
        if ( $mqtt ) {
                print ("MQTT connected to broker " . $cfg->param('mqtt_hostname') . "\n") ;
        } else {
                print ("MQTT ERROR could not connect to broker " . $cfg->param('mqtt_hostname') . "\n") ;
        }

} else {
        print ("MQTT disabled by configuration \n");
}



# Let's go ....

#Present menu of nodes in database and get user to choose one
$nodeId = menu() ;

# Retrieve the configuration and put into mqtt message
$mqttData{msg} = printNodeConfig($nodeId) ;

# Set MQTT topic to emonhub/tx/<nodeid>/values  for TX by emonpi
$mqttTopic =  $mqttTopic . "/$nodeId/values" ;

# If we were able to create the mqtt value string send it
if ($mqttData{msg} eq "0" ) {
	print "ERROR creating configuration - exiting\n" ;
} elsif ($mqttData{msg} eq "1" ) {
	print "Cancelled.\n";
} else {
	# Add the crc 
	$mqttData{msg} .= "," . calcCrc($mqttData{msg}) ;
	HomeAutomation::mqttSend(\$mqtt, $mqttTopic, %mqttData) ;
	$mqtt->tick(5);
	print "Configuration sent to blind - it will move slightly up and down if received OK\n" ;
} 

sleep (2) ;
$dbh->disconnect;
exit () ;
